src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1672,6 +1671,39 @@ public override BoundNode VisitCall(BoundCall node)
 
             return builder == null ? default : builder.ToImmutableAndFree();
 
+            AttributeAnnotations removeInapplicableAnnotations(ParameterSymbol parameter, AttributeAnnotations annotations)
+            {
+                // Ignore NotNullWhenTrue that is inapplicable
+                annotations = removeInapplicableNotNullWhenSense(parameter, annotations, sense: true);
+
+                // Ignore NotNullWhenFalse that is inapplicable
+                annotations = removeInapplicableNotNullWhenSense(parameter, annotations, sense: false);
+
+                // EnsuresTrue must be applied to a bool-returning parameter
+                if ((annotations & AttributeAnnotations.EnsuresTrue) != 0 &&
+                    (parameter.Type?.SpecialType != SpecialType.System_Boolean))
+                {
+                    annotations &= ~AttributeAnnotations.EnsuresTrue;
+                }
+
+                // EnsuresFalse must be applied to a bool-returning parameter
+                if ((annotations & AttributeAnnotations.EnsuresFalse) != 0 &&
+                    (parameter.Type?.SpecialType != SpecialType.System_Boolean))
+                {
+                    annotations &= ~AttributeAnnotations.EnsuresFalse;
+                }
+
+                // We'll ignore EnsuresTrue and EnsuresFalse if both set in metadata
+                if ((annotations & AttributeAnnotations.EnsuresTrue) != 0 &&
+                    (annotations & AttributeAnnotations.EnsuresFalse) != 0)
+                {
+                    annotations &= ~AttributeAnnotations.EnsuresTrue;
+                    annotations &= ~AttributeAnnotations.EnsuresFalse;
+                }
```

Comment:
Perhaps `if ((annotations & Both) == Both) annotations &= Both;`

Reply:
I left this one as-is. The enum doesn't have a value for Both and creating a const here doesn't seem to simplify the code.

Reply:
Ok. Changing this after all, since keep getting the same feedback ;-)

---
In reply to: [192452866](https://github.com/dotnet/roslyn/pull/27313#discussion_r192452866) [](ancestors = 192452866)